File: src/claw_mem/security/checkpoint.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Checkpoint Manager (MVP Version)

Creates regular memory snapshots with rollback support.
"""

import json
from datetime import datetime
from typing import Dict, Optional
from pathlib import Path


class CheckpointManager:
    """
    Checkpoint Manager
    
    MVP version creates simple snapshots. Incremental checks and auto-rollback will be added in future iterations.
    """
    
    def __init__(self, workspace: Path):
        """
        Initialize Checkpoint Manager
        
        Args:
            workspace: Workspace path
        """
        self.workspace = workspace
        self.checkpoint_dir = workspace / ".checkpoints"
        
        # Ensure directory exists
        self.checkpoint_dir.mkdir(parents=True, exist_ok=True)
    
    def create(self, session_id: str) -> str:
        """
        Create checkpoint
        
        Args:
            session_id: Session ID
            
        Returns:
            str: Checkpoint ID
        """
        checkpoint_id = f"{session_id}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        checkpoint_data = {
            "checkpoint_id": checkpoint_id,
            "session_id": session_id,
            "timestamp": datetime.now().isoformat(),
            "status": "created"
        }
        
        # Save checkpoint
        file_path = self.checkpoint_dir / f"{checkpoint_id}.json"
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(checkpoint_data, f, indent=2)
        
        return checkpoint_id
    
    def save(self, session_id: str) -> bool:
        """
        Save session checkpoint
        
        Args:
            session_id: Session ID
            
        Returns:
            bool: Success status
        """
        # MVP version: just log
        checkpoint_id = f"{session_id}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        checkpoint_data = {
            "checkpoint_id": checkpoint_id,
            "session_id": session_id,
            "timestamp": datetime.now().isoformat(),
            "status": "saved"
        }
        
        file_path = self.checkpoint_dir / f"{checkpoint_id}.json"
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(checkpoint_data, f, indent=2)
        
        return True
    
    def rollback(self, checkpoint_id: str) -> bool:
        """
        Rollback to specific checkpoint
        
        Args:
            checkpoint_id: Checkpoint ID
            
        Returns:
            bool: Success status
        """
        # MVP version: not implemented
        print(f"⚠️  Rollback feature not implemented yet: {checkpoint_id}")
        return False
    
    def list_checkpoints(self, session_id: Optional[str] = None) -> list:
        """
        List checkpoints
        
        Args:
            session_id: Session ID filter (optional)
            
        Returns:
            list: Checkpoint list
        """
        checkpoints = []
        
        for file_path in self.checkpoint_dir.glob("*.json"):
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
                if session_id is None or data.get("session_id") == session_id:
                    checkpoints.append(data)
        
        return sorted(checkpoints, key=lambda x: x.get("timestamp", ""), reverse=True)

File: src/security/checkpoint.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Checkpoint Manager (MVP Version)
 *
 * Creates regular memory snapshots with rollback support.
 */

import * as fs from "fs";
import * as path from "path";

export interface CheckpointData {
  checkpoint_id: string;
  session_id: string;
  timestamp: string;
  status: string;
}

export class CheckpointManager {
  private workspace: string;
  private checkpointDir: string;

  /**
   * @param workspace - Workspace directory path
   */
  constructor(workspace: string) {
    this.workspace = workspace;
    this.checkpointDir = path.join(workspace, ".checkpoints");
    fs.mkdirSync(this.checkpointDir, { recursive: true });
  }

  /**
   * Create a checkpoint for a session.
   *
   * @param sessionId - Session ID
   * @returns The checkpoint ID
   */
  create(sessionId: string): string {
    const now = new Date();
    const ts = now
      .toISOString()
      .replace(/[-:]/g, "")
      .replace(/\.\d{3}/, "")
      .replace("T", "_")
      .slice(0, 15);
    const checkpointId = `${sessionId}_${ts}`;

    const data: CheckpointData = {
      checkpoint_id: checkpointId,
      session_id: sessionId,
      timestamp: now.toISOString(),
      status: "created",
    };

    const filePath = path.join(this.checkpointDir, `${checkpointId}.json`);
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
    return checkpointId;
  }

  /**
   * Save a session checkpoint.
   *
   * @param sessionId - Session ID
   * @returns Success status
   */
  save(sessionId: string): boolean {
    const now = new Date();
    const ts = now
      .toISOString()
      .replace(/[-:]/g, "")
      .replace(/\.\d{3}/, "")
      .replace("T", "_")
      .slice(0, 15);
    const checkpointId = `${sessionId}_${ts}`;

    const data: CheckpointData = {
      checkpoint_id: checkpointId,
      session_id: sessionId,
      timestamp: now.toISOString(),
      status: "saved",
    };

    const filePath = path.join(this.checkpointDir, `${checkpointId}.json`);
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
    return true;
  }

  /**
   * Rollback to a specific checkpoint.
   *
   * MVP version: not implemented.
   *
   * @param checkpointId - Checkpoint ID
   * @returns Success status (always false in MVP)
   */
  rollback(checkpointId: string): boolean {
    console.warn(`Rollback feature not implemented yet: ${checkpointId}`);
    return false;
  }

  /**
   * List all checkpoints, optionally filtered by session ID.
   *
   * @param sessionId - Optional session ID filter
   * @returns Sorted list of checkpoint data (newest first)
   */
  listCheckpoints(sessionId?: string): CheckpointData[] {
    const checkpoints: CheckpointData[] = [];

    if (!fs.existsSync(this.checkpointDir)) {
      return checkpoints;
    }

    const files = fs.readdirSync(this.checkpointDir);
    for (const file of files) {
      if (!file.endsWith(".json")) continue;
      const filePath = path.join(this.checkpointDir, file);
      try {
        const raw = fs.readFileSync(filePath, "utf-8");
        const data: CheckpointData = JSON.parse(raw);
        if (sessionId === undefined || data.session_id === sessionId) {
          checkpoints.push(data);
        }
      } catch {
        // skip malformed files
      }
    }

    checkpoints.sort(
      (a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime(),
    );
    return checkpoints;
  }
}
